solutions/coci-21-sjeckanje.mdx

```diff
@@ -0,0 +1,144 @@
+---
+id: coci-21-sjeckanje
+source: COCI 2021
+title: Sjeckanje
+author: Timothy Gao
+---
+
+For any segment $[l, r]$ with more than three elements, if there exists a tuple such that $a_i < a_{i+2} < a_{i+1}$ or $a_i > a_{i+2} > a_{i+1}$ where $i...i+2$ are contained in the segment, then the segment can be split into two segments $l, l+1... i+1$ and $i+2, i+3, ... r$. This is provably at least as good as the score ($\text {max} - \text {min}$) of the initial segment. The tuples $a_i < a_{i+2} < a_{i+1}$ and $a_i > a_{i+2} > a_{i+1}$ must exist in any segment that is **not** non-increasing or non-decreasing, and the first two elements of the segment determines whether the rest of the segment will be non-increasing or non-decreasing. Therefore, an optimal solution can be constructed with only non-increasing or non-decreasing segments. This paves the way for several key properties.
+
+Define array $D$ of length $N-1$ such that $D_i = A_{i+1} - A_i$. If the current segment $s = [l, r]$ is non-decreasing, then $\sum_{i=l}^{r-1} D_i = \text {max}(s) - \text {min}(s)$ and $D_i \geq 0$ for $i \in [l, r]$. If the current segment $s = [l, r]$ is non-decreasing, then $-1 \cdot \sum_{i=l}^{r-1} D_i = \text {max}(s) - \text {min}(s)$ and $D_i \leq 0$ for $i \in [l, r]$. Between any two adjacent selected segments $s = [a, b]$ and $w = [b+1, c]$, $D_b$ must not included in our answer. With this formulation, adding integer $x$ to $l, l+1, ... r$ is equivalent to performing the point updates $D_{l-1}$ += $x$ and $D_{r}$ -= $x$.
+
+Now, the problem becomes selecting segments on $D$ rather than the original $A$, where each segment contributes the absolute value of the sum of its elements to the final answer. Every two neighboring segments must be separated by at least one element (see $D_b$ above), and all elements in each segment must be either $\leq 0$ or $\geq 0$. One way to go from here is with a modified segment tree. Each node stores the maximum value of its segment for each of the four cases of taking or not taking the first or last element, as well as the border elements of its segment.  When merging, we must pay attention to only combine two segments together if the signs of their border elements to be combined are the same.
```

Comment:
```suggestion
Now, the problem becomes selecting segments on $D$ rather than the original $A$, where each segment contributes the absolute value of the sum of its elements to the final answer. Every two neighboring segments must be separated by at least one element (see $D_b$ above), and all elements in each segment must be either $\leq 0$ or $\geq 0$. One way to go from here is with a segment tree. Each node stores the maximum value of its segment for each of the four cases of taking or not taking the first or last element, as well as the border elements of its segment.  When merging, we must pay attention to only combine two segments together if the signs of their border elements to be combined are the same.
```